### ISSUE TITLE
Unexpectedly report error on correct code: 'field' is declared but its value is never read.(6133)'.

### ISSUE THREAD

**Issue by vczh** [microsoft/TypeScript#35153]
**TypeScript Version:**  3.7.2

**Search Terms:** Put the error message but I don't recognize any similar issue by reading titles.

**Code**

```ts
export class A {
    private field: {} = {};

    public constructor() {
        // empty
    }

    public setField(value: {}): void {
        this.setFieldInternal('field', value);
    }

    private setFieldInternal(name: 'field', value: {}): void {
        this[name] = value;
    }
}
```

**Expected behavior:** No error is reported

**Actual behavior:** 'field' is declared but its value is never read.(6133).

Private member `field` **could be assigned** by assigning to `this[name]` while `name` has the type `'field'` (or any string literal union that contains `'field'`). So I believe this error message is incorrect.

If I add `readonly` to this field, `this[name] = value;` will report another error: Cannot assign to 'field' because it is a read-only property.

The compiler leaves me no space for writing code like this.

**Playground Link:** N/A

**Related Issues:** N/A

**Comment by ajafff** [comment]
Are you sure this comes from the compiler? This looks more like a linter error to me.

**Comment by dragomirtitian** [comment]
@ajafff  It does come form the compiler if `noUnusedLocals` is used.

[Playground Link](https://www.typescriptlang.org/play/?noUnusedLocals=true#code/KYDwDg9gTgLgBAYwDYEMDOa4EE4G8CwAUHCXGFAJYBuKMwcAZhcEgCYBceAvnALzcBuIkVJkArgCMkFBIggA7NDChiEMaAAoAlHhGjSAegNxgAWzAwAnnpJdhxUmEnTZaYDABizNhppIxwJy4XFqcVBAUrLoO+nAwABYUaAB0bp7erACS8nRQ8ihIGgDkTCysRQA0cH4BWkIxcHaENmSUNHRwaV5l2bn5hfmmgXAlGZXVBQFBIWERUQQNoglJANqDwAC6fBP+wPWiTVxAA)

**Comment by ajafff** [comment]
@dragomirtitian the OP mentions `Private member variable 'field' is never reassigned; mark it as 'readonly'.` and that's definitely not a compiler error.

The error you get with `noUnusedLocals` (`'field' is declared but its value is never read.`) is correct, because the property is only written and never read.

**Comment by RyanCavanaugh** [comment]
One of these errors comes from tslint

**Comment by vczh** [comment]
I don't think so, since [https://www.typescriptlang.org/play/](https://www.typescriptlang.org/play/) gives me this error too. Unless it has tslint turned on.

**Comment by ajafff** [comment]
@vczh Look at the error message. It's different in the TypeScript playground

**Comment by vczh** [comment]
@ajafff playground shows `'field' is declared but its value is never read.(6133)`, there is still an error. OK let me change the title of this issue.

**Comment by dragomirtitian** [comment]
@vczh  Yes but as @ajafff  pointed out, that error is actually correct in the context of `noUnusedLocals`. Add a read to the field and the error goes away as expected.

[Playground Link](https://www.typescriptlang.org/play/?noUnusedLocals=true#code/KYDwDg9gTgLgBAYwDYEMDOa4EE4G8CwAUHCXGFAJYBuKMwcAZhcEgCYBceAvnALzcBuIkVJkArgCMkFBIggA7NDChiEMaAAoAlHhGjSAegNxgAWzAwAnnpJdhxUmEnTZaYDABizNhppIxwJy4XFqcVBAUrLoO+nAwABYUaAB0bp7erACS8nRQ8ihIGgDkTCysRQA0cH4BWkIxcHaENmSUNHRwaV5l2bn5hfmmgXAlGZXVBQFBIWERUQQNoglJANqDwAC6fBP+wPWxJAgKaBBIwMlIEADmGstoayhDG1otTVxAA)
